net/node.go

```diff
@@ -101,8 +102,11 @@ func NewNode(
 
 	fin := finalizer.NewFinalizer()
 
+	ctx, cancel := context.WithCancel(ctx)
+
 	peerstore, err := pstoreds.NewPeerstore(ctx, db.Peerstore(), pstoreds.DefaultOpts())
 	if err != nil {
+		cancel()
```

Comment:
suggestion: instead of calling `cancel` on every return you could name the return args `(nodeResult *Node, err error)` and call
```go
defer func() {
  if resultNode == nil {
    cancel()
  }
}
```

And then in the end `nodeResult = n`

Reply:
I try to avoid named return variables as much as possible and I don't mind the explicitness of calling cancel wherever it's needed. I'll think about it while I wait the the rest of John's review.

Reply:
I also don't like named returns, but this method has quite some occurrences of `return`. So another person might add some other error check later and forget to add `cancel`.